File: web/src/lib/pathfind/route.ts
import { dist, neighbors } from "./graph";
import { goalProvince } from "./snap";
import type { Leg, PathGraph, Target } from "./types";

type ANode = { id: string; f: number; g: number; parent?: string; index: number; closed?: boolean };

function astar(g: PathGraph, start: string, goal: string): string[] | null {
  const open: ANode[] = [];
  const nodes = new Map<string, ANode>();
  const push = (n: ANode) => {
    n.index = open.length;
    open.push(n);
  };
  const pop = (): ANode | undefined => {
    open.sort((a, b) => a.f - b.f);
    return open.shift();
  };

  const startN: ANode = { id: start, g: 0, f: dist(g, start, goal), index: -1 };
  nodes.set(start, startN);
  push(startN);

  while (open.length > 0) {
    const cur = pop()!;
    if (cur.closed) continue;
    cur.closed = true;
    if (cur.id === goal) {
      const out: string[] = [goal];
      let c = goal;
      for (;;) {
        const n = nodes.get(c);
        if (!n?.parent) break;
        c = n.parent;
        out.push(c);
      }
      return out.reverse();
    }
    for (const nb of neighbors(g, cur.id)) {
      const tentG = cur.g + dist(g, cur.id, nb);
      let nn = nodes.get(nb);
      if (!nn) {
        nn = { id: nb, g: tentG, f: tentG + dist(g, nb, goal), index: -1 };
        nodes.set(nb, nn);
        push(nn);
        continue;
      }
      if (nn.closed) continue;
      if (tentG >= nn.g) continue;
      nn.parent = cur.id;
      nn.g = tentG;
      nn.f = tentG + dist(g, nb, goal);
    }
  }
  return null;
}

function nearestProvince(x: number, y: number, g: PathGraph): string {
  let best = "";
  let bestD = Infinity;
  for (const [id, c] of Object.entries(g.coords)) {
    const d = Math.hypot(x - c.x, y - c.y);
    if (d < bestD) {
      bestD = d;
      best = id;
    }
  }
  return best;
}

// resolveStartOnEdge constrains the routing start province to one of the two
// known edge endpoints when a unit is mid-flight, preventing A* from starting
// at a geometrically closer but topologically unconnected province.
function resolveStartOnEdge(
  g: PathGraph,
  x: number,
  y: number,
  edgeFrom?: string,
  edgeTo?: string,
): string {
  if (!edgeFrom || !edgeTo || edgeFrom === edgeTo) {
    return nearestProvince(x, y, g);
  }
  const coordA = g.coords[edgeFrom];
  const coordB = g.coords[edgeTo];
  if (!coordA || !coordB) {
    return nearestProvince(x, y, g);
  }
  const distA = Math.hypot(x - coordA.x, y - coordA.y);
  const distB = Math.hypot(x - coordB.x, y - coordB.y);
  return distA <= distB ? edgeFrom : edgeTo;
}

export function route(
  g: PathGraph,
  fromX: number,
  fromY: number,
  target: Target,
  edgeFrom?: string,
  edgeTo?: string,
): Leg[] | null {
  const startProv = resolveStartOnEdge(g, fromX, fromY, edgeFrom, edgeTo);
  const goalProv = goalProvince(target);
  if (!startProv || !goalProv) return null;

  let via: string[];
  if (startProv !== goalProv) {
    const path = astar(g, startProv, goalProv);
    if (!path) return null;
    via = path;
  } else {
    via = [startProv];
  }

  const legs: Leg[] = [];
  let curX = fromX;
  let curY = fromY;
  let curProv = startProv;

  for (let i = 1; i < via.length; i++) {
    const next = via[i];
    const c = g.coords[next];
    legs.push({
      fromProv: curProv,
      toProv: next,
      fromX: curX,
      fromY: curY,
      toX: c.x,
      toY: c.y,
    });
    curX = c.x;
    curY = c.y;
    curProv = next;
  }

  const tx = target.x;
  const ty = target.y;
  if (legs.length === 0 || legs[legs.length - 1].toX !== tx || legs[legs.length - 1].toY !== ty) {
    legs.push({
      fromProv: curProv,
      toProv: goalProv,
      fromX: curX,
      fromY: curY,
      toX: tx,
      toY: ty,
    });
  }
  if (legs.length === 0) {
    legs.push({
      fromProv: startProv,
      toProv: goalProv,
      fromX: fromX,
      fromY: fromY,
      toX: tx,
      toY: ty,
    });
  }
  return legs;
}

File: internal/domain/pathdom/route.go
package pathdom

import (
	"container/heap"
	"errors"
	"math"
)

// Leg is one straight segment along the movement graph.
type Leg struct {
	FromProv, ToProv string
	FromX, FromY     float64
	ToX, ToY         float64
}

var ErrNoRoute = errors.New("no route to target")

// resolveStartOnEdge picks the start province for routing when a unit is
// mid-flight on a known edge. It constrains the start to one of the two
// edge endpoints (whichever is closer) rather than the globally nearest
// province center, preventing routes from jumping to unconnected provinces.
func (g *Graph) resolveStartOnEdge(x, y float64, edgeFrom, edgeTo string) string {
	if edgeFrom == "" || edgeTo == "" || edgeFrom == edgeTo {
		return g.nearestProvinceID(x, y)
	}
	coordA, okA := g.coords[edgeFrom]
	coordB, okB := g.coords[edgeTo]
	if !okA || !okB {
		return g.nearestProvinceID(x, y)
	}
	distA := math.Hypot(x-coordA.x, y-coordA.y)
	distB := math.Hypot(x-coordB.x, y-coordB.y)
	if distA <= distB {
		return edgeFrom
	}
	return edgeTo
}

// RouteFromEdge plans a path from (fromX,fromY) to target using A*, where the
// unit is known to be on the edge between edgeFrom and edgeTo. This prevents
// routing from an incorrect province when the unit's position is geometrically
// closer to a third province that is not on the traversed edge.
func (g *Graph) RouteFromEdge(fromX, fromY float64, edgeFrom, edgeTo string, target Target) ([]Leg, error) {
	startProv := g.resolveStartOnEdge(fromX, fromY, edgeFrom, edgeTo)
	return g.routeFromProvince(fromX, fromY, startProv, target)
}

// Route plans a path from (fromX,fromY) to target using A* on provinces.
func (g *Graph) Route(fromX, fromY float64, target Target) ([]Leg, error) {
	startProv := g.nearestProvinceID(fromX, fromY)
	return g.routeFromProvince(fromX, fromY, startProv, target)
}

func (g *Graph) routeFromProvince(fromX, fromY float64, startProv string, target Target) ([]Leg, error) {
	goalProv := target.goalProvince()

	if startProv == "" || goalProv == "" {
		return nil, ErrNoRoute
	}

	var via []string
	if startProv != goalProv {
		path, ok := g.astar(startProv, goalProv)
		if !ok {
			return nil, ErrNoRoute
		}
		via = path
	} else {
		via = []string{startProv}
	}

	legs := make([]Leg, 0, len(via))
	curX, curY := fromX, fromY
	curProv := startProv

	for i := 1; i < len(via); i++ {
		next := via[i]
		c, _ := g.coord(next)
		legs = append(legs, Leg{
			FromProv: curProv, ToProv: next,
			FromX: curX, FromY: curY, ToX: c.x, ToY: c.y,
		})
		curX, curY = c.x, c.y
		curProv = next
	}

	// Final leg to exact target coordinates.
	tx, ty := target.X, target.Y
	if len(legs) == 0 || legs[len(legs)-1].ToX != tx || legs[len(legs)-1].ToY != ty {
		legs = append(legs, Leg{
			FromProv: curProv, ToProv: goalProv,
			FromX: curX, FromY: curY, ToX: tx, ToY: ty,
		})
	}
	if len(legs) == 0 {
		legs = append(legs, Leg{
			FromProv: startProv, ToProv: goalProv,
			FromX: fromX, FromY: fromY, ToX: tx, ToY: ty,
		})
	}
	return legs, nil
}

func (t Target) goalProvince() string {
	if t.Kind == TargetNode {
		return t.Province
	}
	if t.T <= 0.5 {
		return t.EdgeFrom
	}
	return t.EdgeTo
}

type node struct {
	id       string
	f, g      float64
	parent   string
	index    int
	closed   bool
}

type pq []*node

func (h pq) Len() int           { return len(h) }
func (h pq) Less(i, j int) bool { return h[i].f < h[j].f }
func (h pq) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
	h[i].index = i
	h[j].index = j
}
func (h *pq) Push(x any) {
	n := x.(*node)
	n.index = len(*h)
	*h = append(*h, n)
}
func (h *pq) Pop() any {
	old := *h
	n := len(old) - 1
	item := old[n]
	old[n] = nil
	item.index = -1
	*h = old[:n]
	return item
}

func (g *Graph) astar(start, goal string) ([]string, bool) {
	open := &pq{}
	heap.Init(open)

	nodes := map[string]*node{start: {id: start, g: 0, f: g.dist(start, goal), index: -1}}
	heap.Push(open, nodes[start])

	for open.Len() > 0 {
		cur := heap.Pop(open).(*node)
		if cur.closed {
			continue
		}
		cur.closed = true
		if cur.id == goal {
			return reconstruct(nodes, goal), true
		}
		for _, nb := range g.m.Neighbors(cur.id) {
			tentG := cur.g + g.dist(cur.id, nb)
			nn, seen := nodes[nb]
			if !seen {
				nn = &node{id: nb, index: -1}
				nodes[nb] = nn
			}
			if nn.closed {
				continue
			}
			if seen && tentG >= nn.g {
				continue
			}
			nn.parent = cur.id
			nn.g = tentG
			nn.f = tentG + g.dist(nb, goal)
			if nn.index < 0 {
				heap.Push(open, nn)
			} else {
				heap.Fix(open, nn.index)
			}
		}
	}
	return nil, false
}

func reconstruct(nodes map[string]*node, goal string) []string {
	out := []string{goal}
	for cur := goal; ; {
		n := nodes[cur]
		if n == nil || n.parent == "" {
			break
		}
		cur = n.parent
		out = append(out, cur)
	}
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}
	return out
}

// LegLength returns the euclidean length of a leg.
func LegLength(l Leg) float64 {
	return math.Hypot(l.ToX-l.FromX, l.ToY-l.FromY)
}
